IGNITE-25348 Java thin: add partition-aware batch splitting

ascherbakoff opened the pull request:
1. Implemented per-partition batch splitting for explicit txns on a client.
2. Each batch is applied in parallel.
3. Fixed a critical bug causing no-op direct mapping not to enlist in client transaction.
4. Fixed a bug causing empty entries stay in client tx inflight map.

The bug (3) occurs if a first direct enlistment to the partition is no-op operation. This causes enlistment not to be processed on tx finish, so acquired locks are not released until timeout.
The fix involves enlisting on any request outcome and adding some compatibility checks.

ptupitsyn commented:
> Fixed a critical bug causing no-op direct mapping not to enlist in client transaction.
> Fixed a bug causing empty entries stay in client tx inflight map.

@ascherbakoff please create tickets for both of those, and let's have separate PRs so we can review and manage fixes in isolation.

ascherbakoff commented on modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java:
```diff
@@ -877,12 +886,148 @@ int tryGetPartitionCount() {
+                    Map<Integer, List<E>> mapped = new HashMap<>();
+                    for (E key : keys) {
+                        ClientSchema schema = schemaFut.getNow(null);
+                        @Nullable List<String> aff = partitionsFut.getNow(null);
```
🆗

ascherbakoff commented on modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java:
```diff
@@ -877,12 +886,148 @@ int tryGetPartitionCount() {
+                        ClientSchema schema = schemaFut.getNow(null);
+                        @Nullable List<String> aff = partitionsFut.getNow(null);
+                        int hash = hashFunc.apply(schema, key);
+                        Integer part = aff == null ? null : Math.abs(hash % aff.size());
```
Good catch. Simplified the splitting code.

ascherbakoff commented on modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java:
```diff
@@ -877,12 +886,148 @@ int tryGetPartitionCount() {
+                        ClientSchema schema = schemaFut.getNow(null);
+                        @Nullable List<String> aff = partitionsFut.getNow(null);
+                        int hash = hashFunc.apply(schema, key);
+                        Integer part = aff == null ? null : Math.abs(hash % aff.size());
```
🆗

ascherbakoff commented:
> > 3. Fixed a critical bug causing no-op direct mapping not to enlist in client transaction.
> > 4. Fixed a bug causing empty entries stay in client tx inflight map.
> 
> @ascherbakoff please create tickets for both of those and let's have separate PRs so we can review and manage fixes in isolation.

@ptupitsyn I agree in general, but let's keep it as is this time for the sake of simplicity.  I've added detailed bug description to PR head.

ptupitsyn commented on modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java:
```diff
@@ -877,12 +877,135 @@ int tryGetPartitionCount() {
+                    }
+
+                    for (E key : keys) {
+                        ClientSchema schema = schemaFut.getNow(null);
```
Can be moved out of the loop.

ascherbakoff replied:
🆗

ptupitsyn commented on modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java:
```diff
@@ -877,12 +877,135 @@ int tryGetPartitionCount() {
+
+                    int idx = 0;
+                    for (E key : keys) {
+                        ClientSchema schema = schemaFut.getNow(null);
```
Can be moved out of the loop.

ascherbakoff replied:
🆗

ptupitsyn commented on modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java:
```diff
@@ -877,12 +877,135 @@ int tryGetPartitionCount() {
+        return CompletableFuture.allOf(schemaFut, partitionsFut)
+                .thenCompose(v -> {
+                    @Nullable List<String> aff = partitionsFut.getNow(null);
+                    Map<Integer, List<E>> mapped = new HashMap<>();
+                    List<CompletableFuture<R>> res = new ArrayList<>();
```
let's move below null check and specify capacity:

```suggestion
                    Map<Integer, List<E>> mapped = new HashMap<>(aff.size());
                    List<CompletableFuture<R>> res = new ArrayList<>(aff.size());
```
